pydantic/dataclasses.py

```diff
@@ -33,136 +76,210 @@ def __get_validators__(cls: Type['Dataclass']) -> 'CallableGenerator':
+    validate_on_init: Optional[bool] = None,
+) -> Union[Callable[[Type[Any]], 'DataclassClassOrWrapper'], 'DataclassClassOrWrapper']:
+    """
+    Like the python standard lib dataclasses but with type validation.
+    The result is either pydantic dataclass that will validate input data
+    or a wrapper that will trigger validation around a stdlib dataclass
+    to avoid modifying it directly
+    """
+    the_config = get_config(config)
+
+    def wrap(cls: Type[Any]) -> 'DataclassClassOrWrapper':
         import dataclasses
 
-        return cls(**dataclasses.asdict(v))
-    else:
-        raise DataclassTypeError(class_name=cls.__name__)
+        if is_builtin_dataclass(cls):
+            should_validate_on_init = False if validate_on_init is None else validate_on_init
+            _add_pydantic_validation_attributes(cls, the_config, should_validate_on_init, '')
+            return DataclassProxy(cls)
 
+        else:
+            dc_cls_doc = cls.__doc__ or ''  # needs to be done before generating dataclass
+            dc_cls = dataclasses.dataclass(  # type: ignore
+                cls, init=init, repr=repr, eq=eq, order=order, unsafe_hash=unsafe_hash, frozen=frozen
+            )
+            should_validate_on_init = True if validate_on_init is None else validate_on_init
+            _add_pydantic_validation_attributes(dc_cls, the_config, should_validate_on_init, dc_cls_doc)
+            return dc_cls
 
-def _get_validators(cls: Type['Dataclass']) -> 'CallableGenerator':
-    yield cls.__validate__
+    if _cls is None:
+        return wrap
 
+    return wrap(_cls)
 
-def setattr_validate_assignment(self: 'Dataclass', name: str, value: Any) -> None:
-    if self.__initialised__:
-        d = dict(self.__dict__)
-        d.pop(name, None)
-        known_field = self.__pydantic_model__.__fields__.get(name, None)
-        if known_field:
-            value, error_ = known_field.validate(value, d, loc=name, cls=self.__class__)
-            if error_:
-                raise ValidationError([error_], self.__class__)
 
-    object.__setattr__(self, name, value)
+@contextmanager
+def set_validation(cls: Type['DataclassT'], value: bool) -> Generator[Type['DataclassT'], None, None]:
+    original_run_validation = cls.__pydantic_run_validation__
+    try:
+        cls.__pydantic_run_validation__ = value
+        yield cls
+    finally:
+        cls.__pydantic_run_validation__ = original_run_validation
 
 
-def is_builtin_dataclass(_cls: Type[Any]) -> bool:
+class DataclassProxy:
+    __slots__ = '__dataclass__'
+
+    def __init__(self, dc_cls: Type['Dataclass']) -> None:
+        object.__setattr__(self, '__dataclass__', dc_cls)
+
+    def __call__(self, *args: Any, **kwargs: Any) -> Any:
+        with set_validation(self.__dataclass__, True):
+            return self.__dataclass__(*args, **kwargs)
+
+    def __getattr__(self, name: str) -> Any:
+        return getattr(self.__dataclass__, name)
+
```

Comment:
```suggestion

    def __instancecheck__(self, instance):
        return isinstance(instance, self.__dataclass__)

```

Should fix the `assert not isinstance(validated_a, A_PY)` case in the above comment.